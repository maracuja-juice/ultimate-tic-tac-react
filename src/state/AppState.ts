import { Point } from '../util';

export interface AppState {
  currentGame: GameInformation;
  finishedGames: FinishedGameState[];
  analysisGame: AnalysisGame;
}

export interface GameInformation {
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
}

export interface FinishedGameState {
  id?: string;
  date: string;
  winner: Winner;
  gameState: SmallBoardInformation[];
  moves: MoveState[];
  saveState: string;
  errorMessage: string;
}

export interface AnalysisGame {
  id: string;
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
  currentMove: number;
}

export interface GameState {
  currentPlayer: Player;
  isFinished: boolean;
  winningPlayer: Winner;
}

export interface TileInformation {
  position: Point;
  value: TileValue;
}

export interface SmallBoardInformation extends TileInformation {
  tiles: SmallTileInformation[];
}

export interface SmallTileInformation extends TileInformation {
  boardPosition: Point;
}

export enum TileValue {
  Cross = 0,
  Circle = 1,
  Empty = 2,
  Destroyed = 3,
}

export enum Winner {
  Cross = 0,
  Circle = 1,
  None = 2,
  Draw = 3,
}

export enum Player {
  Cross = 0,
  Circle = 1,
}

export interface Move {
  boardPosition: Point;
  tilePosition: Point;
}

export interface MoveState extends Move {
  player: Player;
  moveNumber: number;
}

export interface GenericAction {
  type: string;
  // tslint:disable-next-line: no-any
  payload?: any;
}

export interface MarkSpecially {
  condition: boolean;
  position?: { boardPosition: Point; tilePosition: Point };
}
