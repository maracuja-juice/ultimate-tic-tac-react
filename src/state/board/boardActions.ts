import { TileValue } from '../AppState';
import { Point } from '../../util/Point';

export const SET_TILE_VALUE = 'tile-reducer/set-tile-value';
export const SET_BOARD_VALUE = 'tile-reducer/set-board-value';
export const CALCULATE_BOARD_VALUE = 'tile-reducer/calculate-board-value';

export const setTileValue = ( boardPosition: Point, tilePosition: Point, tileValue: TileValue ) => ({
    type: SET_TILE_VALUE,
    payload: {boardPosition: boardPosition, tilePosition: tilePosition, tileValue: tileValue}
});

export const setBoardValue = ( boardPosition: Point, tileValue: TileValue ) => ({
    type: SET_BOARD_VALUE,
    payload: {boardPosition: boardPosition, tileValue}
});

export const calculateBoardValue = ( boardPosition: Point ) => ({
    type: CALCULATE_BOARD_VALUE,
    payload: boardPosition
});