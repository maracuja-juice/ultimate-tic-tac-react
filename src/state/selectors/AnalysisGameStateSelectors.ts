import { AnalysisGame, AppState, Player } from '../AppState';

export const getAnalysisGame = (
  state: AppState,
  id: string,
): AnalysisGame | undefined => {
  const finishedGame = state.finishedGames.find((g) => g.id === id);
  if (finishedGame) {
    return {
      id: finishedGame.id!,
      board: finishedGame.gameState,
      moves: finishedGame.moves,
      currentMove: finishedGame.moves.length,
      game: {
        currentPlayer: Player.Circle,
        winningPlayer: finishedGame.winner,
        isFinished: true,
      },
      activeBoards: [],
    };
  }

  return undefined;
};
