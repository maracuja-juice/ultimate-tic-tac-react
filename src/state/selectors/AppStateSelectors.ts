import { AnalysisGame, AppState, Player } from '../AppState';

export const getBoards = (state: AppState) => state.currentGame.board;

export const getCurrentPlayer = (state: AppState) =>
  state.currentGame.game.currentPlayer;

export const getWinningPlayer = (state: AppState) =>
  state.currentGame.game.winningPlayer;

export const getMoves = (state: AppState) => state.currentGame.moves;

export const getFinishedGame = (
  state: AppState,
  id: string,
): AnalysisGame | undefined => {
  const finishedGame = state.finishedGames.find((g) => g.id === id);
  if (finishedGame) {
    return {
      moves: finishedGame.moves,
      currentMove: finishedGame.moves.length,
      game: {
        currentPlayer: Player.Circle,
        winningPlayer: finishedGame.winner,
        isFinished: true,
      },
      activeBoards: [],
      board: finishedGame.gameState,
      id: finishedGame.id!,
    };
  }

  return undefined;
};
