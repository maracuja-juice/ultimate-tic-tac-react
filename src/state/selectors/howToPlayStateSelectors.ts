import { createSelector } from 'reselect';
import { getHowToPlay } from './appStateSelectors';
import { HowToPlay } from '../AppState';
import { howToPlaySteps } from '../howToPlay/howToPlaySteps';

export const getCurrentHowToPlayText = createSelector(
  [getHowToPlay],
  (howToPlay: HowToPlay) => {
    return howToPlaySteps[howToPlay.stepNumber].text;
  },
);
