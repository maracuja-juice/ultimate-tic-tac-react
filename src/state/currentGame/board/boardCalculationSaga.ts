import { put, select, takeEvery } from 'redux-saga/effects';
import { GenericAction } from '../../AppState';
import { CALCULATE_BOARD_VALUE, setBoardValue } from './boardActions';
import { playerToTileValue, TicTacToeGame } from '../../../lib';
import { getMoves } from '../../selectors/appStateSelectors';
import { SagaIterator } from 'redux-saga';

function* calculateWinningBoard(action: GenericAction): SagaIterator {
  const boardPosition = action.payload;

  const moves = yield select(getMoves);

  const game = new TicTacToeGame(moves);
  const winResult = game.getWinResultForSmallBoard(boardPosition);
  if (winResult.isFinished) {
    const newSmallBoardTileValue = playerToTileValue(
      winResult.winningPlayer,
      true,
    );
    yield put(setBoardValue(boardPosition, newSmallBoardTileValue));
  }
}

function* boardCalculationSaga() {
  yield takeEvery(CALCULATE_BOARD_VALUE, calculateWinningBoard);
}

export default boardCalculationSaga;
