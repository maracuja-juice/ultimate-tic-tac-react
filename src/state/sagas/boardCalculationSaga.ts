import { put, select, takeEvery } from 'redux-saga/effects';
import { GenericAction, SmallBoardInformation } from '../AppState';
import { CALCULATE_BOARD_VALUE, setBoardValue } from '../board/boardActions';
import { getWinResult } from '../../util/CheckBoard';
import { arePointsEqual, playerToTileValue } from '../../util';
import { getBoards } from '../selectors/AppStateSelectors';

function* calculateWinningBoard( action: GenericAction ) {
    const boardPosition = action.payload;

    const boards = yield select( getBoards );

    const affectedBoard = boards.find(
        ( board: SmallBoardInformation ) =>
            arePointsEqual( board.position, boardPosition )
    ).tiles;

    const winResult = getWinResult( affectedBoard );
    if (winResult.isFinished) {
        const newSmallBoardTileValue = playerToTileValue( winResult.winningPlayer!, true );
        yield put( setBoardValue( boardPosition, newSmallBoardTileValue ) );
    }
}

function* boardCalculationSaga() {
    yield takeEvery( CALCULATE_BOARD_VALUE, calculateWinningBoard );
}

export default boardCalculationSaga;