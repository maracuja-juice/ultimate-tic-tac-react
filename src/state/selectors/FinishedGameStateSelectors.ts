import { createSelector } from 'reselect';
import { getBoards, getIsReplay, getMoves, getWinningPlayer } from './AppStateSelectors';

export const getWinningPlayerAsString = createSelector( [getWinningPlayer], (winningPlayer => {
    if (winningPlayer === 0) {
        return 'X';
    } else if (winningPlayer === 1) {
        return 'Y';
    } else if (winningPlayer === null) {
        return null;
    } else {
        return undefined;
    }
}) );

export const getFinishedGameData = createSelector(
    [getWinningPlayerAsString, getIsReplay, getBoards, getMoves],
    ( winningPlayer, isReplay, boards, moves ) => {
        return {
            winner: winningPlayer,
            gameState: boards,
            isReplay,
            moves
        };
    } );