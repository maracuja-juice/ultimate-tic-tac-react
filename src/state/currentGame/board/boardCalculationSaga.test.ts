import { expectSaga } from 'redux-saga-test-plan';
import boardCalculationSaga from './boardCalculationSaga';
import { getMoves } from '../../selectors/AppStateSelectors';
import { select } from 'redux-saga/effects';
import { CALCULATE_BOARD_VALUE, SET_BOARD_VALUE } from './boardActions';
import { Player } from '../../AppState';
import { movesForUnfinishedBoardMock } from '../../../mocks';

describe('boardCalculationSaga', () => {
  it('should dispatch the set board value for the board with the Cross Player', () => {
    return expectSaga(boardCalculationSaga)
      .provide([[select(getMoves), movesForUnfinishedBoardMock]])
      .put({
        type: SET_BOARD_VALUE,
        payload: { boardPosition: { x: 0, y: 1 }, tileValue: Player.Cross },
      })
      .dispatch({ type: CALCULATE_BOARD_VALUE, payload: { x: 0, y: 1 } })
      .silentRun();
  });

  it('should dispatch the set board value for the board with the Circle Player', () => {
    return expectSaga(boardCalculationSaga)
      .provide([[select(getMoves), movesForUnfinishedBoardMock]])
      .put({
        type: SET_BOARD_VALUE,
        payload: { boardPosition: { x: 1, y: 1 }, tileValue: Player.Circle },
      })
      .dispatch({ type: CALCULATE_BOARD_VALUE, payload: { x: 1, y: 1 } })
      .silentRun();
  });

  // if more put effects happen: this catches it + this checks for the order
  it('should match snapshot', () => {
    return expectSaga(boardCalculationSaga)
      .provide([[select(getMoves), movesForUnfinishedBoardMock]])
      .dispatch({ type: CALCULATE_BOARD_VALUE, payload: { x: 0, y: 0 } })
      .silentRun()
      .then((result) => {
        expect(result.toJSON()).toMatchSnapshot();
      });
  });
});
