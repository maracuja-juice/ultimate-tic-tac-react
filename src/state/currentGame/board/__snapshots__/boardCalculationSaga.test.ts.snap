// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`boardCalculationSaga should match snapshot 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "tile-reducer/calculate-board-value",
          "@@redux-saga-test-plan/json/function/calculateWinningBoard",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeEvery",
      },
      "type": "FORK",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {
            "payload": Object {
              "x": 0,
              "y": 0,
            },
            "type": "tile-reducer/calculate-board-value",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/calculateWinningBoard",
      },
      "type": "FORK",
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "payload": Object {
            "boardPosition": Object {
              "x": 0,
              "y": 0,
            },
            "tileValue": 0,
          },
          "type": "tile-reducer/set-board-value",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/getMoves",
      },
      "type": "SELECT",
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "tile-reducer/calculate-board-value",
      },
      "type": "TAKE",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "tile-reducer/calculate-board-value",
      },
      "type": "TAKE",
    },
  ],
}
`;
